### PR TITLE
Add address_of(array[index]) support to incremental SMT2 backend

### PR DESCRIPTION
**Pull request by esteffin**
Add `address_of(array[index])` support to new incremental SMT2 backend.
Enabled all `cbmc-primitives` regressions tests working with new SMT2 backend.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <solvers/smt2_incremental/smt_core_theory.h>
 
 #include <functional>
+#include <iostream>
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :no_entry_sign: I am assuming this is debugging code which should not be included in this PR.

File: src/solvers/smt2_incremental/convert_expr_to_smt.h [hunk 2]
```diff
@@ -16,6 +16,11 @@ class typet;
 ///   stored as sort ast (abstract syntax tree).
 smt_sortt convert_type_to_smt_sort(const typet &type);
 
+/// \brief Lower the `address_of(array[idx])` sub expressions in \p to
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.h]: :pick: `\p` should be followed by the name of the parameter.

File: src/solvers/smt2_incremental/convert_expr_to_smt.h [hunk 3]
```diff
@@ -16,6 +16,11 @@ class typet;
...
+///   `idx + address_of(array)`, so that it can be feed to
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.h]: :pick: `feed` -> `fed`.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 4]
```diff
@@ -1813,6 +1814,28 @@ at_scope_exitt<functiont> at_scope_exit(functiont exit_function)
...
+  expr.visit_pre([](exprt &expr) {
+    if(
+      const auto address_of_expr =
+        expr_try_dynamic_cast<address_of_exprt>(expr))
+    {
+      if(
+        const auto array_index_expr =
+          expr_try_dynamic_cast<index_exprt>(address_of_expr->object()))
+      {
+        expr = plus_exprt{
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :bulb: I think deeply nested control flow can be tricky to follow. Therefore I suggest refactoring this based on listing the rejection criteria rather than nesting the success criteria.
```
    const auto address_of_expr = expr_try_dynamic_cast<address_of_exprt>(expr);
    if(!address_of_expr)
      return;
    const auto array_index_expr =
      expr_try_dynamic_cast<index_exprt>(address_of_expr->object());
    if(!array_index_expr)
      return;
    expr = plus_exprt{
      address_of_exprt{
        array_index_expr->array(),
        type_checked_cast<pointer_typet>(address_of_expr->type())},
      array_index_expr->index()};
```

**Review by thomasspriggs — COMMENTED** [review]
The only thing this really needs doing is removing the debug include of `iostream`. Therefore I am happy to approve, once that comment is addressed.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 1 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Done

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 4 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Seems a good idea.
Changed.

File: src/solvers/smt2_incremental/convert_expr_to_smt.h (hunk 3 above)
Comment (esteffin) [reply src/solvers/smt2_incremental/convert_expr_to_smt.h]: Corrected

**Review by esteffin — COMMENTED** [review]
Changed as required.

**Review by thomasspriggs — APPROVED** [review]
Looks good to me.

**Review by TGWDB — APPROVED** [review]
